### DISCUSSION TITLE
Poetry vs. PyScaffold: Pick One?

### DISCUSSION THREAD

**Discussion by stratofax** [ion-oset/developer-commons#17] [Infrastructure]
As a result of my research for the wiki page [Python: Create a Project with Poetry in GitHub](https://github.com/ion-oset/developer-commons/wiki/Python:-Create-a-Project-with-Poetry-in-GitHub), I've realized that there's a significant overlap in what these two tools aspire to do do. Both have the ability to set up a Python project, creating the files and directory structures needed for a PyPi-compliant Python package.

The main difference, as far as I can tell, is this:

* Poetry is an "all-in-one" solution for project setup, package management, virtual environment, and package building. Poetry uses the pyproject.toml file **exclusively** to track project settings.
* PyScaffold is focused on setting up a project by installing all the tools you need for package management (pip or pipenv), virtual environments, and package building (with tox). 

In other words, PyScaffold installs the **other** tools you need and then gets out of the way, whereas Poetry does all of these tasks from creating the project directory structure to building the package and publishing it. As a result, projects created with PyScaffold have many more files than a Poetry-built project.

This begs the question: is it really a good idea to use both Poetry and PyScaffold when we could use one or the other? 

By using both, we introduce conflicts (both tools want to write to `pyproject.toml`, for example). We also introduce confusion and ambiguity. For example, should I use `tox` to build my package, or `poetry build`?

I suggest that we select **one** of these tools for OSET Python projects. Based on the application of "Occam's razor for coding" (pick the simplest solution that gets the job done now) I'd recommend that we go with Poetry, because it does everything we need it to do, and we reduce the number of tools in the toolkit by a literal order of magnitude.

What do you think?

**Comment by cwulfman** [comment]
I like Pyscaffold because it generates directories and templates for you, but that's about it.  I think @ion-oset is right in suggesting we can do that ourselves with github templates, or Cookiecutter.

**Reply by stratofax** [reply]
Actually, Poetry can do that for you too:

```bash
poetry new --src my-package
```
Creates the following directory structure:

```
my-package
├── pyproject.toml
├── README.rst
├── src
│   └── my_package
│       └── __init__.py
└── tests
    ├── __init__.py
    └── test_my_package.py
```

Notice the use of the `--src` flag to put the package code dir inside a source dir; you can also use the `--name` flag to specify a different name for your package name than the project name.

Details here: [Commands: new | Documentation | Poetry](https://python-poetry.org/docs/cli/#new)

**Comment by stratofax** [comment]
I wanted to create a new Python repo for a package I'm writing focused on PathLib, and I thought: "this would be a great test case for using Poetry to do it all!" So I wrote up the process of setting up a Python project in Poetry and then adding the repo to GitHub. Here's what I've got so far: [Python: Create a project with Poetry and add it to GitHub](https://github.com/ion-oset/developer-commons/wiki/Python:-Create-a-project-with-Poetry-and-add-it-to-GitHub).

It's still a draft, I need to add more links and a section on adding and installing packages, but this worked as advertised!

**Reply by ion-oset** [reply]
Can I suggest moving this post to be under #16 (Create a Python Project with Poetry in GitHub) instead? It seems like a better fit there.

**Comment by ion-oset** [comment]
> I've realized that there's a significant overlap in what these two tools aspire to do do. Both have the ability to set up a Python project, creating the files and directory structures needed for a PyPi-compliant Python package.

I don't see the overlap between PyScaffold and Poetry as significant. They can both be used to initialize a project but they are not trying to do the same thing. They can work together and I'll explain a little about how. But I'm going to do that in a reply to this post because I think this comparison is asking the wrong question.

We aren't going to have a single kind of Python project, or even a single kind of language project. There will be libraries, standalone CLI tools, API servers, configuration files for cloud-hosting and so on. It's likely that no tool is configurable enough to cover all our use cases. More importantly we will want to capture best practice that's subtle that no tool _should_ be expected to. If there are specific needs we have for e.g. a FastAPI Python scaffold for a common set of web services, but another scaffold is needed for Java compatibility we need to have both. We might be able to use e.g. Cookie Cutter or some other general templating tool to get our projects started but after that we will want to configure and tweak, and I believe we won't want to have to constantly update instructions for how to make a new project to match.

Instead of asking people who work on ElectOS projects to run a tool, I believe the instructions should boil down to: take one of a set of carefully curated project scaffolds and copy it. (A natural way to do this is to set up [GitHub template projects](https://docs.github.com/en/repositories/creating-and-managing-repositories/creating-a-template-repository), and have people fork the right one.) Working on a new service analogous to a Versa or Vanadium? Fork `electos-fastapi-service` (for Python) or whatever equivalent we have for other environments. As we discover we need to update the templates we make incremental updates to the scaffolds/templates as we need them, and people can merge in those changes.

If we want to start with a template made by Poetry great.  But that's not a decision we should be locked into or dependent on.

**Comment by ion-oset** [comment]
Responding to the issues with the specific tools. Quoted replies are to the [original post](https://github.com/ion-oset/developer-commons/discussions/17#discussion-3918111).

**TL;DR**: We're going to want to mix tools that overlap and we'll have to figure out on a case by case basis. It's possible to mix PyScaffold and Poetry if we want to, with some editing. It's definitely possible to mix Poetry and Tox. `pyproject.toml` is designed to be a shared file, with namespaced sections for different tools. The only place they can collide is if they want to set the `build_system` section.

I personally want Poetry to be our _build system_, but there are uses for the other tools that are not precluded by using Poetry to do that.

Specifics in replies to this post.

**Reply by ion-oset** [reply]
### PyScaffold and Poetry

If one wants to make a new project with PyScaffold and maintain it with Poetry this is possible. PyScaffold is more flexible than Poetry for template generation (Poetry is deliberately limited in its choices there). Poetry is primarily a build system though. The overlap is minor.

It's true that used without guidance they trip over each other, but they are occupying different phases of the development process. PyScaffold creates a project directory and fills it out. It's used once and then the scaffold is modified to suit. The project it creates knows nothing about Poetry, but there's no reason it couldn't be modified to be Poetry compatible. After that PyScaffold would only be used if we need to make a fundamental change to the scaffold. Poetry on the other hand continues to be used over and over, because it's primary purpose is to manage dependencies and build the project.

> Both tools want to write to `pyproject.toml`, for example.

That's not unique to these toosl. In practice `pyproject.toml` holds configuration for _all_ Python tools. For Poetry that's a build spec, but it also is used for configurations for _tox_ or for _black_ and so on. The file will always take some careful management and version control. But  `pyproject.toml` is broken up into sections, and each tool is supposed to claim it's own namespace. The only fundamental incompatibility between what the PyScaffold and Poetry generate is one section of `pyproject.toml`:

PyScaffold:

```
[build-system]
requires = ["setuptools>=46.1.0", "setuptools_scm[toml]>=5", "wheel"]
build-backend = "setuptools.build_meta"
```
Poetry:

```
[build-system]
requires = ["poetry-core>=1.0.0"]
build-backend = "poetry.core.masonry.api"
```

There are redundant files generated, but those can be cleaned up. If we don't require a specific tool, we do this rarely, and developers only need to stay synced up with the scaffold repository.

**Reply by ion-oset** [reply]
### Tox and Poetry

> We also introduce confusion and ambiguity. For example, should I use tox to build my package, or poetry build

This is a different question than with PyScaffold, as both Poetry and Tox drive builds and it's better to have a default. But there are situations where we want a regular build (Poetry) and need the advanced dependency management, and ones where we need to test on all support Python versions (Tox). We can end up needing to use both and to resolve the proper steps regardless.

In this case there are established ways to do that. While it's a little subtle Poetry and Tox are intended to play well with each other:

- [Poetry: Is Tox supported?](https://python-poetry.org/docs/faq/#is-tox-supported)
- [Tox: Example with Poetry](https://tox.wiki/en/stable/example/package.html#poetry)

Also:
- [How can I get tox and poetry to work together to support testing multiple versions of a Python dependency?](https://stackoverflow.com/questions/59377071/#59522588) (Stack Overflow)
- [Python: A toxic work environment (tox & poetry)](https://levelup.gitconnected.com/python-a-toxic-work-environment-tox-poetry-b8200563c474) (Despite the poorly chosen title it's about how _well_ Poetry and Tox work together).

There are also tools to connect the two (though I haven't used them):

- [tox-poetry-installer](https://pypi.org/project/tox-poetry-installer/)
- [tox-poetry](https://pypi.org/project/tox-poetry/)

**Comment by cwulfman** [comment]
I just came across this discussion with the Pyscaffold developer re Poetry integration: https://github.com/pyscaffold/pyscaffoldext-pyproject/issues/2.  By the end of the thread the developer seems to think it is feasible; what seems to hold things for the other dev is the lack of support for automatic versioning (which setuptools provides).

I'm not advocating for anything, btw.  Whatever makes sense for ElectOS software is fine with me.

**Reply by stratofax** [reply]
I read this thread too; unfortunately things seemed to have tapered off before the integration tool being discussed was implemented. It does describe a real-world scenario where developers are using both PyScaffold and Poetry together quite  productively, in a manner that seems similar to what we're considering, especially the use of a custom PyScaffold extension to support different project types.
